web-animations/timing-model/animations/updating-the-finished-state.html

```diff
@@ -0,0 +1,131 @@
+<script src="/resources/testharnessreport.js"></script>
+<script src="../../testcommon.js"></script>
+<body>
+<script>
+'use strict';
+
+function validateFinishEvent(t, event, animation, expectedEventTime) {
+  t.step(function() {
+    assert_equals(event.target, animation, "Animation should be event target");
+    assert_times_equal(event.currentTime, expectedEventTime,
+        "Event currentTime should be the given expected value");
+    assert_times_equal(event.timelineTime, animation.startTime + expectedEventTime,
+        "Event timelineTime should be the animation's start time plus the given expected value");
+  });
+}
+
+function awaitFinishEventAndPromise(t, animation, expectedEventTime) {
+  var eventPromise = new Promise(function(resolve) {
+    animation.onfinish = function(event) {
+      validateFinishEvent(t, event, animation, expectedEventTime);
+      resolve();
+    };
+  });
+
+  var finishedPromise = animation.finished.then(function(target) {
+    assert_equals(target, animation);
+  });
+
+  var timeout = waitForAnimationFrames(3).then(function() {
+    return Promise.reject("Expected finish notifications did not occur");
+  });
+
+  return Promise.race([timeout,Promise.all([eventPromise, finishedPromise])]);
+}
+
+promise_test(function(t) {
+  var div = createDiv(t);
+  var animation = div.animate(null, 1);
+  animation.onfinish = t.unreached_func("Seeking to finish should not fire finish event");
+  animation.finished.then(t.unreached_func("Seeking to finish should not resolve finished promise"));
+  animation.currentTime = 1;
+  animation.currentTime = 0;
+  animation.pause();
+  return waitForAnimationFrames(3);
+}, "Finish notification steps don't run when the animation seeks to finish and then seeks back again");
+
+promise_test(function(t) {
+  var div = createDiv(t);
+  var animation = div.animate(null, 1);
+  animation.onfinish = t.unreached_func("Seeking past finish should not fire finish event");
+  animation.finished.then(t.unreached_func("Seeking past finish should not resolve finished promise"));
+  animation.currentTime = 10;
+  animation.currentTime = 0;
+  animation.pause();
+  return waitForAnimationFrames(3);
+}, "Finish notification steps don't run when the animation seeks past finish and then seeks back again");
+
+promise_test(function(t) {
+  var div = createDiv(t);
+  var animation = div.animate(null, 1);
+  animation.finish();
+  animation.currentTime = 0;
+  animation.pause();
+  var expectedEventTime = 1;
+  return awaitFinishEventAndPromise(t, animation, expectedEventTime);
+}, "Finish notification steps run when the animation completes with .finish(), even if we then seek away");
+
+promise_test(function(t) {
+  var div = createDiv(t);
+  var animation = div.animate(null, 1);
+  animation.currentTime = 10;
+  var expectedEventTime = 10;
+  return awaitFinishEventAndPromise(t, animation, expectedEventTime);
+}, "Finish notification steps run when the animation seeks past finish");
```

Comment:
I don't understand how this is supposed to work with the way `validateFinishEvent` is defined.

Here's what will happen in Firefox at least
1. `animation` is initially pending
1. We update the current time on it, but since it has a hold time, we just update that
1. We run the "update an animation's finished state" procedure but we don't update the hold time since we have a pending task, and we don't run finish notification steps because 'current finished state' is pending.
1. Sometime between this frame and the next we resolve the start time by recording the current timeline time (but at this point the timeline time is not locked to the frame time--this is possibly where we differ)
1. On the next frame we'll update the start time with the time we recorded and then go to run the finish steps. However, the timeline time is now not necessarily equal to `startTime + 10`. In fact, even if we do lock start times to frame times, I don't really understand why `timelineTime` would necessarily equal `startTime + 10`.

What is Chrome doing here and why is this condition supposed to hold?

`event.timelineTime` is going to be bound to a frame time but I don't think `animation.startTime` is. But perhaps that's not even the issue.

This animation is currently in its pending state, now suppose that sometime in between the current frame and the next frame the UA finishes getting the animation ready and records its start time. There are probably two possible options here:
a) `animation.startTime` = next frame time (does Chrome do this?)
b) `animation.startTime` = some time between this frame time and the next frame

Then on the next frame we'll probably still be at current time < 10s (definitely for (a), probably for (b)).
On the next frame we'll